### PR TITLE
REALMC-7586: Include new sync types (UUID, Mixed) and expose information to determine container types (Dictionary, Set)

### PR DESCRIPTION
**Pull request by tkaye407**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 *.sw[nop]
 *.iml
 .vscode/
+.idea/
```
Comment (tkaye407) [inline .gitignore:4]: This was annoying me, so I added it in here

File: schema.go [hunk 2]
```diff
@@ -169,7 +169,7 @@ func (d *Schema) parseSchema(documentNode interface{}, currentSchema *subSchema)
 					ErrorDetails{"x": KEY_DEFINITIONS, "y": TYPE_OBJECT},
 				))
 			}
-			
+
```
Comment (tkaye407) [inline schema.go:172]: gofmt fixed this

File: schemaPool.go [hunk 3]
```diff
@@ -168,6 +194,9 @@ func (p *schemaPool) GetDocument(reference gojsonreference.JsonReference) (*sche
 	refToURL.GetUrl().Fragment = ""
 
 	if cachedSpd, ok := p.schemaPoolDocuments[refToURL.String()]; ok {
+		if bsonD, isBsonD := cachedSpd.Document.(bson.D); isBsonD {
```
Comment (tkaye407) [inline schemaPool.go]: This fixes an issue where the tests were failing here: https://github.com/10gen/gojsonschema/blob/9b64c19c895bb7d38d2526ab22303b2bc8510f46/jsonschema_test.go#L1174

File: subSchema.go [hunk 4]
```diff
@@ -190,6 +190,27 @@ func (s *SubSchemaAccessor) Title() *string {
 	return s.schema.title
 }
 
+func (s *SubSchemaAccessor) UniqueItems() bool {
```
Comment (tkaye407) [inline subSchema.go]: This exposes what we need to determine if this is a Set / Dictionary

File: schemaPool.go (hunk 3 above)
Comment (joshmadewell) [inline schemaPool.go]: nit: it's more idiomatic to use `ok` here I think? shouldn't affect the `ok` on 196 either.

File: subSchema.go (hunk 4 above)
Comment (joshmadewell) [inline subSchema.go]: it's totally fine if you don't want to do this but is there anyway I can get you to move all of `SubSchemaAccessor` out to it's own file? It should help reduce merge conflicts in the future when we try to update this file with upstream.

File: subSchema.go [hunk 5]
```diff
@@ -190,6 +190,27 @@ func (s *SubSchemaAccessor) Title() *string {
...
+	return s.schema.uniqueItems
+}
+
+func (s *SubSchemaAccessor) AdditionalProperties() []string {
```
Comment (joshmadewell) [inline subSchema.go]: This accessor doesn't make much sense to me. `additionalProperties` can only be a `boolean` or an `object` but we're returning that it's a `[]string`. This looks like business logic more than something I'd expect to be in the library.

Comment (joshmadewell) [reply subSchema.go]: https://json-schema.org/understanding-json-schema/reference/object.html#id2

**Review by joshmadewell — CHANGES_REQUESTED** [review]

File: subSchema.go (hunk 5 above)
Comment (tkaye407) [reply subSchema.go]: Changed this to return (bool, SubSchemaAccessor)

**Comment by tkaye407** [comment]
@joshmadewell I see you are working tomorrow but then heading out for a bit, let me know if you have time to take another look at this before heading out as it blocks the rest of the DataType support work

File: schemaPool.go [hunk 6]
```diff
@@ -136,6 +137,31 @@ func (p *schemaPool) parseReferencesRecursive(document interface{}, ref gojsonre
 	return nil
 }
 
+// bsonDToMap converts a bson.D to a bson.M at all nesting levels
+// we cannot use the bson.D.Map() function because that is only one level
+func bsonDToMap(doc bson.D) map[string]interface{} {
+	newDoc := map[string]interface{}{}
+	for _, entry := range doc {
+		if subDoc, isSubDoc := entry.Value.(bson.D); isSubDoc {
```
Comment (joshmadewell) [inline schemaPool.go]: nit: using `ok` throughout this function in these type assertions -- the reason for `ok` can be implied by the code itself, the variable name seems unnecessary :D

**Review by joshmadewell — APPROVED** [review]
got one more comment re: `ok` vs `isBlegh` but otherwise seems good. would like to see one person from sync review this before we merge just to 👍 that we're implementing the expected spec but nothing egregious here imo

**Review by mpobrien — APPROVED** [review]

**Pull request merged**